sound/soc/sof/intel/hda-loader.c

```diff
@@ -372,3 +372,148 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
 	dev_err(sdev->dev, "error: load fw failed err: %d\n", ret);
 	return ret;
 }
+
+// skl/kbl enable core and code loader DMA has some difference with apl/cnl
+// add the APIs for skl/kbl
+static int cl_stream_prepare_skl(struct snd_sof_dev *sdev, unsigned int format,
+				 unsigned int size, struct snd_dma_buffer *dmab,
+				 int direction)
+{
+	u32 tag = 0;
+	return tag;
+}
+
+static int cl_dsp_init_skl(struct snd_sof_dev *sdev, const void *fwdata,
+			   u32 fwsize)
+{
+	int tag, ret, i;
+	u32 hipcie;
+	u32 reg;
+	const struct sof_intel_dsp_desc *chip = sdev->hda->desc;
+
+	if (hda_dsp_core_is_enabled(sdev, HDA_DSP_CORE_MASK(0))) {
+		ret = hda_dsp_core_stall_reset(sdev, HDA_DSP_CORE_MASK(0));
+		if (ret < 0)
+			goto err;
+
+		ret = hda_dsp_core_run(sdev, HDA_DSP_CORE_MASK(0));
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: dsp core start failed %d\n",
+				ret);
+			ret = -EIO;
+			goto err;
+		}
+	} else {
+		ret = hda_dsp_core_reset_power_down(sdev, HDA_DSP_CORE_MASK(0));
+		if (ret < 0) {
+			dev_err(sdev->dev, "dsp core0 disable fail: %d\n", ret);
+			return ret;
+		}
+		ret = hda_dsp_enable_core(sdev, HDA_DSP_CORE_MASK(0));
+	}
+
+	/* prepare DMA for code loader stream */
+	tag = cl_stream_prepare_skl(sdev, 0x40, fwsize, &sdev->dmab,
+				    SNDRV_PCM_STREAM_PLAYBACK);
+
+	if (tag <= 0) {
+		dev_err(sdev->dev, "error: dma prepare for fw loading err: %x\n",
+			tag);
+		return tag;
+	}
+
+	memcpy(sdev->dmab.area, fwdata, fwsize);
+
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_BAR, HDA_DSP_REG_ADSPIC,
+				HDA_DSP_ADSPIC_IPC, HDA_DSP_ADSPIC_IPC);
+
+	/* enable IPC DONE interrupt */
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_BAR, chip->ipc_ctl,
+				HDA_DSP_REG_HIPCCTL_DONE,
+				HDA_DSP_REG_HIPCCTL_DONE);
+
+	/* enable IPC BUSY interrupt */
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_BAR, chip->ipc_ctl,
+				HDA_DSP_REG_HIPCCTL_BUSY,
+				HDA_DSP_REG_HIPCCTL_BUSY);
+
+	ret = snd_sof_dsp_register_poll(sdev, HDA_DSP_BAR,
+					HDA_DSP_SRAM_REG_ROM_STATUS_SKL,
+					HDA_DSP_ROM_STS_MASK, HDA_DSP_ROM_INIT,
+					HDA_DSP_INIT_TIMEOUT);
+	if (ret >= 0)
+		goto out;
+
+	ret = -EIO;
+
+err:
+	hda_dsp_dump(sdev, SOF_DBG_REGS | SOF_DBG_PCI | SOF_DBG_MBOX);
+	//sdev->dsp_ops.cleanup(sdev->dev, &sdev->dmab, tag);
+	hda_dsp_core_reset_power_down(sdev, HDA_DSP_CORE_MASK(0) |
+				      HDA_DSP_CORE_MASK(1));
+	return ret;
+
+out:
+	return tag;
+}
+
+static int cl_copy_fw_skl(struct snd_sof_dev *sdev, int tag)
+{
+	return -1;
+}
+
+int hda_dsp_cl_boot_firmware_skl(struct snd_sof_dev *sdev)
+{
+	struct snd_sof_pdata *plat_data = dev_get_platdata(sdev->dev);
+	struct firmware stripped_firmware;
+	int ret, tag;
+
+	stripped_firmware.data = plat_data->fw->data;
```

Comment:
You will need to temporarily add an offset here to strip the extended manifest when using the closed source firmware (as it has an extended manifest pre pended ato the start of the FW binary file) otherwise the FW will not authenticate or boot..

Reply:
ask @keyonjie for extended manifest details.

Reply:
yes, FW will not boot without the offset change and remove of extended manifest